main.go

```diff
@@ -31,6 +31,398 @@ const bwSuccessPrefix = "✓"
+		fmt.Fprintln(w, "STEP\tERROR\t")
+		for svc, err := range errorMessages {
+			fmt.Fprintf(w, "%v\t%v\n", svc, err)
+		}
+		w.Flush()
+		return cli.Exit("", 1)
+	}
+
+	return nil
+}
+
+/**
+ * Try to register system against Red Hat Subscription Management server (candlepin server)
+ */
+func registerRHSM(ctx *cli.Context) error {
+	uuid, err := getConsumerUUID()
+	if err != nil {
+		return cli.Exit(err, 1)
+	}
+
+	successPrefix, _, _, isColorful := getColorPreferences(ctx)
+
+	if uuid == "" {
+		username := ctx.String("username")
+		password := ctx.String("password")
+		organization := ctx.String("organization")
+		activationKeys := ctx.StringSlice("activation-key")
+
+		if len(activationKeys) == 0 {
+			if username == "" {
+				password = ""
+				scanner := bufio.NewScanner(os.Stdin)
+				fmt.Print("Username: ")
+				_ = scanner.Scan()
+				username = strings.TrimSpace(scanner.Text())
+			}
+			if password == "" {
+				fmt.Print("Password: ")
+				data, err := term.ReadPassword(int(os.Stdin.Fd()))
+				if err != nil {
+					return cli.Exit(err, 1)
+				}
+				password = string(data)
+				fmt.Printf("\n\n")
+			}
+		}
+
+		var s *spinner.Spinner
+		if isColorful {
+			s = spinner.New(spinner.CharSets[9], 100*time.Millisecond)
+			s.Suffix = " Connecting to Red Hat Subscription Management..."
+			s.Start()
+		}
+		var err error
+		if len(activationKeys) > 0 {
+			fmt.Printf("Registering using activation key")
+			err = registerActivationKey(
+				organization,
+				ctx.StringSlice("activation-key"),
+				ctx.String("server"))
+		} else {
+			if organization != "" {
+				err = registerPassword(username, password, organization, ctx.String("server"))
+			} else {
+				/* TODO: When organization was not specified using CLI option --organization and it is
+				   required, because user is member of more than one organization, then ask for
+				   the organization. RHSM D-Bus API can provide list of available organizations
+				   in this case. */
+				err = registerPassword(username, password, "", ctx.String("server"))
+			}
+		}
+		if isColorful {
+			s.Stop()
+		}
+		if err != nil {
+			return cli.Exit(err, 1)
+		}
+		fmt.Printf(successPrefix + " Connected to Red Hat Subscription Management\n")
+	} else {
+		fmt.Printf(successPrefix + " This system is already connected to Red Hat Subscription Management\n")
+	}
+	return nil
+}
+
+func connectAction(ctx *cli.Context) error {
+	var start time.Time
+	durations := make(map[string]time.Duration)
+	errorMessages := make(map[string]error)
+	hostname, err := os.Hostname()
+	if err != nil {
+		return cli.Exit(err, 1)
+	}
+
+	fmt.Printf("Connecting %v to %v.\nThis might take a few seconds.\n\n", hostname, Provider)
+
+	successPrefix, _, errorPrefix, isColorful := getColorPreferences(ctx)
+
+	/* 1. Register to RHSM, because we need to get consumer certificate. This blocks following action */
+	start = time.Now()
+	/* Note: All successful messages are printed inside registerRHSM() */
+	err = registerRHSM(ctx)
```

Comment:
This doesn't follow the same pattern as steps 2 and 3. Can it be refactored so that all 3 steps follow a similar code pattern?

Reply:
Yeah, I also do not like that the pattern is different in this case. I will try to think about it more and change it to use same pattern in all cases.